src/vue/fields/DateField.vue

```diff
@@ -93,139 +119,226 @@ export default {
+     *
+     * @param {string|date} newValue
+     */
+    value (newValue) {
+      // Prevent updates if v-model value is same as input's current value
+      if (newValue === this.flatpickrDate) return
+      // Sets the current selected date after value changed
+      if (this.fp) this.fp.setDate(newValue, true)
+    },
+    /**
+     * Watch for any config changes and redraw date-picker
+     *
+     * @param {object} newConfig
+     */
+    config: {
+      deep: true,
+      handler (newConfig) {
+        let safeConfig = Object.assign({}, newConfig)
+        // Workaround: Don't pass hooks to configs again otherwise
+        // previously registered hooks will stop working
+        // Notice: we are looping through all events
+        // This also means that new callbacks cannot be passed once a component
+        // has been initialized
+        ALLOWED_FLATPICKR_EVENTS.forEach((hook) => {
+          delete safeConfig[hook]
+        })
+
+        // set new config
+        this.fp.set(safeConfig)
+      },
     },
   },
 
-  created () {
-    this.flatpickrDate = this.value
+  mounted () {
+    // Return early if flatpickr is already mounted
+    if (this.isFlatpickrMounted) return
+
+    let safeConfig = Object.assign({}, this.config)
+
+    // Inject defined methods into events array
+    Object.values(BUILDIN_EVENTS).forEach(hook => {
+      // eslint-disable-next-line
+      safeConfig[hook] = this.arrayify(safeConfig[hook] || [])
+        .concat((...args) => this[hook](...args))
+    })
+
+    // Set initial date without emitting any event
+    safeConfig.defaultDate = this.value || safeConfig.defaultDate
+
+    // Init flatpickr
+    this.fp = new Flatpickr(this.$refs.dateField, safeConfig)
+    this.isFlatpickrMounted = true
+
+    this.flatpickrDate = this.value || safeConfig.defaultDate || null
+  },
+
+  /**
+   * Free up memory
+   */
+  beforeDestroy () {
+    if (this.fp) {
+      this.fp.destroy()
+      this.fp = null
+    }
   },
 
   methods: {
     dateFieldUpdated (event) {
       if (event) {
-        this.$emit('input', this.flatpickrDate)
+        // Let's wait for DOM to be updated
+        this.$nextTick(() => {
+          this.$emit(EMITABLE_EVENTS.input, this.flatpickrDate)
+        })
       }
     },
-    onOpen () {
+    /**
+     * @param {array} selectedDates - an array of Date objects selected by the
+     *        user. When there are no dates selected, the array is empty.
+     * @param {string} dateStr - a string representation of the latest selected
+     *        Date object by the user. The string is formatted as per the
+     *        dateFormat option.
+     * @param {object} instance - the flatpickr object, containing various
+     *        methods and properties.
+     */
+    onOpen (selectedDates, dateStr, instance) {
       this.isCalendarOpen = true
+      // Let's wait for DOM to be updated
+      this.$nextTick(() => {
+        this.$emit(EMITABLE_EVENTS.onOpen)
+      })
     },
```

Comment:
`dateFieldUpdated` is a native method but onOpen and onClose are hooks actually that will be provided concatenated to native flatpickr hooks. `onOpen` and `onClose` meaning is magical, we should not compicate the code by defining methods that actually used like hooks only. We can fix things up like:

```js
methods: {
  hookFlatpickrHooks (fpConfig) {
    const hooks = {
      onOpen: _ => { /* ... */ },
      onClose: _ => { /* ... */ }
    }

    fpConfig[hook] = this.arrayify(safeConfig[hook] || [])
      .concat((...args) => hooks[hook](...args))
  }
}
```
just as example

Reply:
I cannot see any improvements in this solution. Yes, we define the realization of our hooks into a single place, but it will make our code ugly and will make too long nesting in the future. Yes, I understand that we can define hook name in the `BUILTIN_EVENTS` but we can forget to define hook realization and our component will be broken up, but cmon, it's isn't a problem